### ISSUE TITLE
Error range inspection

### ISSUE THREAD

**Issue by bclothier** [rubberduck-vba/Rubberduck#4822]
According to the MS-VBAL specifications, which unfortunately also leaves too much to the reader to fill in the bits, the expected error range is `0-65535`. It does talk about `vbObjectError` which technically is outside the range. However the specification also fail to mention that `0` is apparently handled differently by VBA, and that behavior is different if you try to specify an error number outside of the 2 ranges:

```
1-65535
vbObjectError - vbObjectError + 65535
```

If you try to raise an error with number `0`, >= `65535`, VBA will ignore the number and yield a `5: Invalid procedure call or argument` instead. 

If you try to raise an error with any negative number, VBA may hand back whatever error message it finds for that number, even if it has no relevance to whatever you're trying to do. 

But there's more! If you try to raise an error with `vbObjectError + N`, and fail to define a message, VBA will happily dispense you with error message it may find - [here's a sample](https://docs.microsoft.com/en-us/windows/desktop/com/com-error-codes-1). Note that `vbObjectError` = `0x80040000`.

Additionally, the error number 1 to 512 are reserved by VBA, so defining custom error in that range could be said to be iffy. Defining a custom error in range between 513 to 65535 may overlap with host application's defined error.

Defining a custom error in the range between `vbObjectError` to `vbObjectError` + 65535 makes the most sense but only if you actually set the rest of the parameters (the description, the source and optionally the help file & context). 

Therefore, we need an inspection (or inspections?) that validates:

1) All `Err.Raise N` statements are constrained to a range of from 1 to 65535.
2) All `Err.Raise N, Source, Description`, statements are constrained to a range of `vbObjectError` to `vbObjectError` + 65535. 

Note that this means that statements like `Err.Raise 5, , "My custom message about an invalid action"` would issue the inspection. That could be debatable. I'm inclined to encourage users toward sticking to the `vbObjectError` to `vbObjectError` + 65535 range for the custom errors - after all that's what it is for and discourage overloading the built-in error numbers.

**Comment by retailcoder** [comment]
FWIW I disagree that overloading the built-in error numbers is bad. To me raising error 5 with a custom message is similar to throwing an `ArgumentException` with said custom message, and making a new custom exception type for it makes no sense. If I see error 5, I know it has to do with an invalid argument. If I see "invalid 'foo' argument" with a weird error number, then I immediately think "but why?".

I don't think this can be inspected.... unless I give up throwing error 5. Convince me that my ArgumentException thinking is wrong or evil, and I'll back this up 100% =)

**Comment by bclothier** [comment]
For the record, I am guilty of having the same, overloading the built-in error number which is why I noted in the OP that the part about inspection the built-in error range can be debatable. 

The primary justification I can offer is this - keep in the mind that .NET gives us the ability to create and derive from exception classes, so we can easily create custom exceptions for situations where a custom message isn't enough. That doesn't exist in VBA where we get is a dumb ol' number & 2 strings. This is a problem when you're writing code that is meant to be consumed by other code rather than users directly. If the higher level code is supposed to act up on the error information, customizing the built-in error range forces the higher level code to parse the strings which we all know is very much fun. To avoid that, it'd be easier for the higher level code to parse by numbers only and then decide on how to response. This is easily done in a `Select Case`. 

Now, a counterpoint can be made that such code written in VBA are much more rare; 98%+ of the VBA code will be intended for user consumption in which case we just display the error as-is. In that situation, the error number matters less and the strings matter more. I can see having 2 separate settings for the inspection, with the built-in error range being a hint-level while the other inspection for outside the built-in error except for the `vbObjectError` range being a warning level.

**Comment by retailcoder** [comment]
Agreed, VBA errors aren't exception types. *But* IMO the best practice recommendation to "use the best applicable built-in exception type whenever possible" is solid, and correctly transposes to VBA errors, at least error 5, which I mentally translate into ""invalid operation" or "invalid argument", which I believe are the most common built-in exceptions to throw.

OTOH I can totally see custom errors in the built-in range as being an inspection-worthy problem... maybe have a setting to exclude some specific error numbers?
